src/app/helpers/paginationTemplate/paginationTemplate.ts

```diff
@@ -0,0 +1,198 @@
+import { Directive, OnInit } from "@angular/core";
+import { ActivatedRoute, NavigationEnd, Router } from "@angular/router";
+import { ApiFilter } from "@baw-api/api-common";
+import { ApiErrorDetails } from "@baw-api/api.interceptor.service";
+import {
+  defaultApiPageSize,
+  InnerFilter,
+  Subsets,
+} from "@baw-api/baw-api.service";
+import { PageComponent } from "@helpers/page/pageComponent";
+import { AbstractModel } from "@models/AbstractModel";
+import { NgbPaginationConfig } from "@ng-bootstrap/ng-bootstrap";
+import { noop, Observable, Subject } from "rxjs";
+import { filter, map, mergeMap, takeUntil } from "rxjs/operators";
+
+const queryKey = "query";
+const pageKey = "page";
+
+@Directive()
+// tslint:disable-next-line: directive-class-suffix
+export abstract class PaginationTemplate<I, M extends AbstractModel>
+  extends PageComponent
+  implements OnInit {
+  /**
+   * Observable which updates scroll elements
+   */
+  public apiRequest$ = new Subject<void>();
+  /**
+   * Tracks whether to display the pagination buttons
+   */
+  public displayPagination: boolean;
+  /**
+   * Tracks whether an error has occurred
+   */
+  public error: ApiErrorDetails;
+  /**
+   * Tracks whether an api request is in process
+   */
+  public loading: boolean;
+  /**
+   * Tracks additional filter constraints
+   */
+  public filter: string;
+  /**
+   * Maximum number of elements for current filter
+   */
+  public collectionSize: number;
+  /**
+   * Tracks the current filter page
+   */
+  private _page: number;
+
+  constructor(
+    protected router: Router,
+    protected route: ActivatedRoute,
+    /**
+     * Config for pagination defaults
+     */
+    protected config: NgbPaginationConfig,
+    /**
+     * API Service which will create filter requests
+     */
+    protected api: ApiFilter<M, any[]>,
+    /**
+     * Key to match filter inputs against
+     */
+    private filterKey: keyof I,
+    /**
+     * API Service parameters required to make a filter request
+     */
+    private apiParams: () => any[],
+    /**
+     * Function which will run after each api update (ie. scroll/filter)
+     */
+    private apiUpdate: (models: M[]) => void,
+    /**
+     * Default filter values, may be overridden by later requests
+     */
+    private defaultFilter: InnerFilter<I> = {}
+  ) {
+    super();
+  }
+
+  public ngOnInit() {
+    // Set pagination defaults
+    this.config.maxSize = 3;
+    this.config.pageSize = defaultApiPageSize;
+    this.config.rotate = true;
+    this.displayPagination = false;
+
+    this._page = 1;
+
+    this.apiRequest$
+      .pipe(
+        mergeMap(() => this.getModels()),
```

Comment:
Is `MergeMap` the right choice here?  Maybe `switchMap`?

https://www.learnrxjs.io/learn-rxjs/operators/transformation/switchmap

> Why use switchMap?
The main difference between switchMap and other flattening operators is the cancelling effect. On each emission the previous inner observable (the result of the function you supplied) is cancelled and the new observable is subscribed. You can remember this by the phrase switch to a new observable.
This works perfectly for scenarios like typeaheads where you are no longer concerned with the response of the previous request when a new input arrives. This also is a safe option in situations where a long lived inner observable could cause memory leaks, for instance if you used mergeMap with an interval and forgot to properly dispose of inner subscriptions. Remember, switchMap maintains only one inner subscription at a time, this can be seen clearly in the first example.

Reply:
I think I just grokked this. You need merge map to flatten the multiple observables returned by the filter request?

Which is what bugged me about getModels in the first place...